### PR TITLE
Date/time overhaul

### PR DESCRIPTION
**Pull request by garyb**
Ok, I think this is ready now.

It's a fully PureScript based representation, but we privately hand off some logic to the JS date stuff, for now at least.

I didn't include `Locale` or `Hour12` in here in the end, to keep this library down to just the types needed to represent and manipulate date/times. We will probably want to add some more operations over time too, but I think this is a reasonable starting place.

The other parts of datetime are now in two other libraries - `purescript-now` for fetching the current date, and `purescript-js-date` for the `JSDate` type and general interop. We'll probably also want to introduce a library for convenient parsing and printing of date formats fairly soon too.

### CODE REVIEW

**Comment by paf31** [comment]
I like the split, but I wonder why the dependencies aren't reversed, with the smaller, lower-level `JSDate` library at the bottom? Does that make sense? It seems like it would just be a job of moving the functions which use both types into this library.

**Comment by garyb** [comment]
Eventually I would prefer this library not to use any FFI, so I think the dependencies are better this way around - ideally the datetime library should be backend agnostic, and pulling `js-date` along with it runs counter to that. It would also mean moving the conversion functions into `datetime` rather than having them in `js-date`, which again, rather specialises it to JS use cases.

**Comment by paf31** [comment]
That makes sense. In that case, I'll review this a bit more later tonight.

**Comment by garyb** [comment]
I re-included the locale stuff - it seems reasonable to provide a method of representing local date/time values, and I've updated `purescript-now` with cases where it makes sense to do so: https://github.com/purescript-contrib/purescript-now/blob/bf071638903fa17f036065d190129a41a3411982/src/Control/Monad/Eff/Now.purs#L27-L38. Without `Locale` these might have been confusing, as the returned values would always be in UTC.

File: src/Data/Date.purs [hunk 1]
```diff
@@ -1,298 +1,82 @@
...
-dayOfWeekFromEnum Saturday  = 6
-
-foreign import nowImpl :: forall e. (JSDate -> Date) -> Eff (now :: Now | e) Date
-
-foreign import jsDateConstructor :: forall a. a -> JSDate
-
-foreign import jsDateMethod :: forall a. Fn2 String JSDate a
-
-foreign import strictJsDate :: Fn3 (forall a. a -> Maybe a) (forall a. Maybe a) String (Maybe JSDate)
+  show (Date y m d) = "(Date " <> show y <> " " <> show m <> " " <> show d <> ")"
```
Comment (paf31) [inline src/Data/Date.purs]: Why not use `gShow` here?

File: src/Data/Date/Component.purs [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+-- | Lowers a year value to a plain number.
+unYear :: Year -> Int
+unYear (Year y) = y
+
+derive instance eqYear :: Eq Year
+derive instance ordYear :: Ord Year
+derive instance genericYear :: Generic Year
+
+instance boundedYear :: Bounded Year where
+  bottom = Year (-271820)
```
Comment (paf31) [inline src/Data/Date/Component.purs]: Where do these values come from?

**Comment by paf31** [comment]
Looks great! Sorry for the delayed review.

File: src/Data/Date.purs (hunk 1 above)
Comment (garyb) [inline src/Data/Date.purs]: I could certainly do that, for pretty much all the `Show` instances actually. Is that preferable in this case then?

File: src/Data/Date/Component.purs (hunk 2 above)
Comment (garyb) [inline src/Data/Date/Component.purs]: JavaScript :sob:

The JS Date object supports ±100,000,000 days from the Unix epoch. By using these year bounds we're restricted to slightly less than that, but as long as the month and day are valid, the date will also be. If we used the "actual" `bottom` year, the `bottom` for `Date` could not be implemented with `bottom bottom bottom`, as the 1st of Jan for the earliest supported year is outside the day range. Same for `top`.

Basically it's some arbitrary nonsense that can go away when we implement the date logic in PureScript instead.

**Comment by garyb** [comment]
No problem! It's a pretty big changeset. I'll merge and release this after we decide on whether `gShow` is better to use.

**Comment by garyb** [comment]
I'll add a comment in the code about the seemingly arbitrary year bounds too actually.

File: src/Data/Date.purs (hunk 1 above)
Comment (jdegoes) [inline src/Data/Date.purs]: I don't think core libraries should use the generic implementations because they are much slower; and the libraries don't change very often (at least in theory 😆) so the overhead of manually writing and maintaining the instances is much lower.

**Comment by garyb** [comment]
@paf31 is this good to go then?

**Comment by paf31** [comment]
I think so! 🚢 it.
